calour/ms1_experiment.py

```diff
@@ -196,3 +161,129 @@ def get_bad_features(self, mz_tolerance=0.001, rt_tolerance=2, corr_thresh=0.8,
+        return exp
+
+    def filter_mz_rt(self, mz=None, rt=None, mz_tolerance=0.05, rt_tolerance=0.2, inplace=False, negate=False):
+        '''Filter metabolites based on m/z and/or retention time
+
+        Keep (or remove if negate=True) metabolites that have an m/z and/or retention time close (up to tolerance)
+        to the requested mz and/or rt (or list of mz and/or rt).
+
+        Parameters
+        ----------
+        mz: float or list of float or None, optional
+            the M/Z to filter
+            if None, do not filter based on M/Z
+        rt: float or list of float or None, optional
+            the retention time to filter
+            if None, do not filter based on rt
+        mz_tolerance: float, optional
+            the M/Z tolerance. filter metabolites with abs(metabolite_mz - mz) <= mz_tolerance
+        rt_tolerance: float, optional
+            the rt tolerance. filter metabolites with abs(metabolite_rt - rt) <= rt_tolerance
+        inplace: bool, optional
+            True to replace current experiment, False to create new experiment with results
+        negate: bool, optional
+            If False, keep only metabolites matching mz
+            If True, remove metabolites matching mz
+
+        Returns
+        -------
+        calour.MS1Experiment
+            features filtered based on mz
+        '''
+        if mz is None and rt is None:
+            raise ValueError('at least one of "mz" and "rt" must not be None')
+        if mz is not None:
+            if 'MZ' not in self.feature_metadata.columns:
+                raise ValueError('The Experiment does not contain the column "MZ". cannot filter by mz')
+            else:
+                mz = _to_list(mz)
+        if rt is not None:
+            if 'RT' not in self.feature_metadata.columns:
+                raise ValueError('The Experiment does not contain the column "RT". cannot filter by rt')
+            else:
+                rt = _to_list(rt)
+
+        keep = set()
+        notfound = 0
+        if mz is None:
+            mz = [None] * len(rt)
+        if rt is None:
+            rt = [None] * len(mz)
+        if len(mz) != len(rt):
+            raise ValueError('mz and rt must have same length')
+
+        for cmz, crt in zip(mz, rt):
+            if cmz is not None:
+                mzdiff = np.abs(self.feature_metadata['MZ'] - cmz)
+                keepmz = mzdiff <= mz_tolerance
+            else:
+                keepmz = np.full([len(self.feature_metadata)], True)
+            if crt is not None:
+                rtdiff = np.abs(self.feature_metadata['RT'] - crt)
+                keeprt = rtdiff <= rt_tolerance
+            else:
+                keeprt = np.full([len(self.feature_metadata)], True)
+            bothok = np.logical_and(keepmz, keeprt)
+            if bothok.sum() == 0:
+                notfound += 1
+            keep = keep.union(set(np.where(bothok)[0]))
```

Comment:
as on this for loop, could you use boolean mask array (as I mentioned in your original code of this PR)? you can take a look at `select` variable in filtering.py. It will make the code cleaner and a little more efficient?